src/python/pants/engine/internals/graph.py

```diff
@@ -403,19 +467,13 @@ async def resolve_addresses_with_origins(specs: Specs) -> AddressesWithOrigins:
     )
     # It's possible to resolve the same address both with filesystem specs and address specs. We
     # dedupe, but must go through some ceremony for the equality check because the OriginSpec will
-    # differ. We must also consider that the filesystem spec may have resulted in a generated
-    # subtarget; if the user explicitly specified the original owning target, we should use the
-    # original target rather than its generated subtarget.
+    # differ.
     address_spec_addresses = FrozenOrderedSet(awo.address for awo in from_address_specs)
     return AddressesWithOrigins(
-        [
+        (
             *from_address_specs,
-            *(
-                awo
-                for awo in from_filesystem_specs
-                if awo.address.maybe_convert_to_base_target() not in address_spec_addresses
```

Comment:
Why this change?

Reply:
I've now updated the tests for this, but: subtarget expansion between `UnexpandedTargets` and `Targets` will result in a deduping of all the inputs into just the subtargets, and so doing it here was redundant. It feels like there is a benefit to having fewer special cases in `Specs` -> `Addresses` expansion, to make it closer to a one-in-one-out situation with regard to literals.

Having said that, the deduping of filesystem and address `Specs` is also a special case, so arguably that could be removed as well to allow it to happen during `TargetExpansion` instead.